File: src-tauri/src/commands/frontend/code.rs
use super::utils::*;
use crate::{
    app,
    types::{HexColor, Point, RelativeColoredPoint, RgbOffset},
};

#[tauri::command]
pub fn generate_find_image_code(
    subpath: String,
    start_point: Point,
    end_point: Point,
    threshold: f64,
    min_template_side: u32,
) -> CommandResult<String> {
    let code = app::get().try_with_project(|project| {
        project.interpreter.code().find_image(
            &subpath,
            &start_point,
            &end_point,
            threshold,
            min_template_side,
        )
    })?;
    Ok(code)
}

#[tauri::command]
pub fn generate_find_images_code(
    subpath: String,
    start_point: Point,
    end_point: Point,
    threshold: f64,
    min_template_side: u32,
) -> CommandResult<String> {
    let code = app::get().try_with_project(|project| {
        project.interpreter.code().find_images(
            &subpath,
            &start_point,
            &end_point,
            threshold,
            min_template_side,
        )
    })?;
    Ok(code)
}

#[tauri::command]
pub fn generate_find_relative_colors_code(
    points: Vec<RelativeColoredPoint>,
    start_point: Point,
    end_point: Point,
    rgb_offset: RgbOffset,
) -> CommandResult<String> {
    let code = app::get().try_with_project(|project| {
        project.interpreter.code().find_relative_colors(
            &points,
            &start_point,
            &end_point,
            &rgb_offset,
        )
    })?;
    Ok(code)
}

#[tauri::command]
pub fn generate_find_colors_code(
    hex_colors: Vec<HexColor>,
    start_point: Point,
    end_point: Point,
    rgb_offset: RgbOffset,
) -> CommandResult<String> {
    let code = app::get().try_with_project(|project| {
        project
            .interpreter
            .code()
            .find_colors(&hex_colors, &start_point, &end_point, &rgb_offset)
    })?;
    Ok(code)
}

#[tauri::command]
pub fn generate_recognize_text_code(start_point: Point, end_point: Point) -> CommandResult<String> {
    let code = app::get().try_with_project(|project| {
        project
            .interpreter
            .code()
            .recognize_text(&start_point, &end_point)
    })?;
    Ok(code)
}

#[tauri::command]
pub fn generate_move_absolute_code(point: Point) -> CommandResult<String> {
    let code =
        app::get().try_with_project(|project| project.interpreter.code().absolute_code(&point))?;
    Ok(code)
}

File: src/api/Code.ts
import { invoke } from "@tauri-apps/api/core";
import {
  f64,
  hexColor,
  Point,
  RelativeColoredPoints,
  rgbColor,
  u32,
} from "@types";

class Code {
  static async generateFindImageCode(
    subpath: string,
    startPoint: Point,
    endPoint: Point,
    threshold: f64,
    minTemplateSide: u32,
  ): Promise<string> {
    try {
      return await invoke("generate_find_image_code", {
        subpath,
        startPoint,
        endPoint,
        threshold,
        minTemplateSide,
      });
    } catch (e: unknown) {
      throw e;
    }
  }

  static async generateFindImagesCode(
    subpath: string,
    startPoint: Point,
    endPoint: Point,
    threshold: f64,
    minTemplateSide: u32,
  ): Promise<string> {
    try {
      return await invoke("generate_find_images_code", {
        subpath,
        startPoint,
        endPoint,
        threshold,
        minTemplateSide,
      });
    } catch (e) {
      throw e;
    }
  }

  static async generateFindRelativeColorsCode(
    points: RelativeColoredPoints,
    startPoint: Point,
    endPoint: Point,
    rgbOffset: rgbColor,
  ): Promise<string> {
    try {
      const serializedPoints = points.map((p) => ({
        colored_point: p.coloredPoint.clone(),
        relative_point: p.relativePoint.clone(),
      }));
      return await invoke("generate_find_relative_colors_code", {
        points: serializedPoints,
        startPoint,
        endPoint,
        rgbOffset,
      });
    } catch (e) {
      throw e;
    }
  }

  static async generateFindColorsCode(
    hexColors: hexColor[],
    startPoint: Point,
    endPoint: Point,
    rgbOffset: rgbColor,
  ): Promise<string> {
    try {
      return await invoke("generate_find_colors_code", {
        hexColors,
        startPoint,
        endPoint,
        rgbOffset,
      });
    } catch (e) {
      throw e;
    }
  }

  static async generateRecognizeTextCode(
    startPoint: Point,
    endPoint: Point,
  ): Promise<string> {
    try {
      return await invoke("generate_recognize_text_code", {
        startPoint,
        endPoint,
      });
    } catch (e) {
      throw e;
    }
  }
  static async generateMoveToAbsolutePositionCode(
    point: Point,
  ): Promise<string> {
    try {
      return await invoke("generate_move_absolute_code", {
        point,
      });
    } catch (e) {
      throw e;
    }
  }
}

export { Code };
